src/libpmemfile-posix/data.c

```diff
@@ -0,0 +1,1604 @@
+
+			block = block_list_insert_after(vinode, NULL);
+			block->offset = offset;
+			file_allocate_block_data(pfp, block, count, false);
+			block_cache_insert_block(vinode->blocks, block);
+		} else if (TOID_IS_NULL(block->next)) {
+			/* After the last allocated block */
+
+			block = block_list_insert_after(vinode, block);
+			block->offset = offset;
+			file_allocate_block_data(pfp, block, size, over);
+			block_cache_insert_block(vinode->blocks, block);
+		} else {
+			/* In a hole between two allocated blocks */
+
+			struct pmemfile_block *next = D_RW(block->next);
+
+			/* How many bytes in this hole can be used? */
+			uint64_t hole_count = next->offset - offset;
+
+			/* Are all those bytes needed? */
+			if (hole_count > size)
+				hole_count = size;
+
+			if (hole_count > 0) { /* Is there any hole at all? */
+				block = block_list_insert_after(vinode, block);
+				block->offset = offset;
+				file_allocate_block_data(pfp, block, hole_count,
+				    false);
+				block_cache_insert_block(vinode->blocks, block);
+
+				if (block->size > hole_count)
+					block->size = (uint32_t)hole_count;
+			} else {
+				block = next;
+			}
+		}
+	} while (size > 0);
+}
+
+static struct pmemfile_block *
+find_following_block(struct pmemfile_vinode *vinode,
+	struct pmemfile_block *block)
+{
+	if (block != NULL)
+		return D_RW(block->next);
+	else
+		return vinode->first_block;
+}
+
+enum cpy_direction { read_from_blocks, write_to_blocks };
+
+static void
+read_block_range(const struct pmemfile_block *block,
+	uint64_t offset, uint64_t len, char *buf)
+{
+	ASSERT(len > 0);
+	ASSERT(block == NULL || offset < block->size);
+	ASSERT(block == NULL || offset + len <= block->size);
+
+	/* block == NULL means reading from a hole in a sparse file */
+
+	/*
+	 * !is_block_data_initialized(block) means reading from an
+	 * fallocat-ed region in a file, a region that was allocated,
+	 * but never initialized.
+	 */
+
+	if ((block != NULL) && is_block_data_initialized(block)) {
+		const char *read_from = D_RO(block->data) + offset;
+		memcpy(buf, read_from, len);
+	} else {
+		memset(buf, 0, len);
+	}
+}
+
+static void
+write_block_range(PMEMfilepool *pfp, struct pmemfile_block *block,
+	uint64_t offset, uint64_t len, const char *buf)
+{
+	ASSERT(block != NULL);
+	ASSERT(len > 0);
+	ASSERT(offset < block->size);
+	ASSERT(offset + len <= block->size);
+
+	char *data = D_RW(block->data);
+
+	if (!is_block_data_initialized(block)) {
+		char *start_zero = data;
+		size_t count = offset;
+		if (count != 0) {
+			VALGRIND_ADD_TO_TX(start_zero, count);
+			pmemobj_memset_persist(pfp->pop, start_zero, 0, count);
+			VALGRIND_REMOVE_FROM_TX(start_zero, count);
+		}
+
+		start_zero = data + offset + len;
+		count = block->size - (offset + len);
+		if (count != 0) {
+			VALGRIND_ADD_TO_TX(start_zero, count);
+			pmemobj_memset_persist(pfp->pop, start_zero, 0, count);
+			VALGRIND_REMOVE_FROM_TX(start_zero, count);
+		}
+
+		TX_ADD_FIELD_DIRECT(block, flags);
```

Comment:
Actually this would be:
TX_SET_DIRECT(block, flags, block->flags | BLOCK_INITIALIZED);
which is too long.